Flutter doesn't understand filesystem paths used by gradle plugin 3.0.0

I'm using Android Studio 3 preview 3 and updated my gradle plugin version to 3.0.0-alpha3 and when running flutter run in terminal I get a FileSystemException. The path does exist on my machine so I'm unsure why the error is being generated. I also filed an issue related to running flutter build apk (link below).

https://github.com/flutter/flutter/issues/10629

Flutter crash report; please file at https://github.com/flutter/flutter/issues.

## command

flutter run -d FA71A0302167

## exception

FileSystemException: FileSystemException: Cannot copy file to 'build/app/outputs/apk/app.apk', path = 'build/app/outputs/apk/app-debug.apk' (OS Error: No such file or directory, errno = 2)

```
#0      _File.throwIfError (dart:io/file_impl.dart:599)
#1      _File.copySync (dart:io/file_impl.dart:321)
#2      _LocalFileSystemEntity&ForwardingFile.copySync (package:file/src/forwarding/forwarding_file.dart:25)
#3      buildGradleProjectV2 (package:flutter_tools/src/android/gradle.dart:250)
<asynchronous suspension>
#4      buildGradleProject (package:flutter_tools/src/android/gradle.dart:192)
<asynchronous suspension>
#5      buildApk (package:flutter_tools/src/commands/build_apk.dart:83)
<asynchronous suspension>
#6      AndroidDevice.startApp (package:flutter_tools/src/android/android_device.dart:354)
<asynchronous suspension>
#7      FlutterDevice.runHot (package:flutter_tools/src/resident_runner.dart:233)
<asynchronous suspension>
#8      HotRunner.run (package:flutter_tools/src/run_hot.dart:183)
<asynchronous suspension>
#9      RunCommand.runCommand (package:flutter_tools/src/commands/run.dart:329)
<asynchronous suspension>
#10     FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:210)
<asynchronous suspension>
#11     RunCommand.verifyThenRunCommand (package:flutter_tools/src/commands/run.dart:217)
<asynchronous suspension>
#12     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:156)
<asynchronous suspension>
#13     CommandRunner.runCommand (package:args/command_runner.dart:194)
<asynchronous suspension>
#14     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:262)
<asynchronous suspension>
#15     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:109)
#16     new Future.sync (dart:async/future.dart:206)
#17     CommandRunner.run (package:args/command_runner.dart:109)
#18     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:157)
#19     run.<anonymous closure> (package:flutter_tools/executable.dart:142)
<asynchronous suspension>
#20     AppContext._run (package:flutter_tools/src/base/context.dart:76)
<asynchronous suspension>
#21     AppContext.runInZone.<anonymous closure> (package:flutter_tools/src/base/context.dart:66)
#22     _rootRun (dart:async/zone.dart:1120)
#23     _CustomZone.run (dart:async/zone.dart:1001)
#24     runZoned (dart:async/zone.dart:1467)
#25     AppContext.runInZone (package:flutter_tools/src/base/context.dart:65)
#26     run (package:flutter_tools/executable.dart:114)
<asynchronous suspension>
#27     main (package:flutter_tools/executable.dart:67)
<asynchronous suspension>
#28     main (file:///Users/albertlardizabal/dev/flutter/packages/flutter_tools/bin/flutter_tools.dart:16)
#29     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:263)
#30     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
```

## flutter doctor

```
[✓] Flutter (on Mac OS X 10.12.5 16F73, locale en-US, channel master)
    • Flutter at /Users/albertlardizabal/dev/flutter
    • Framework revision 0f277fcc8a (6 days ago), 2017-06-06 17:19:27 +0200
    • Engine revision 8686a45810
    • Tools Dart version 1.24.0-dev.3.0

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/albertlardizabal/Library/Android/sdk
    • Platform android-25, build-tools 25.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[-] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b
    ✗ libimobiledevice and ideviceinstaller are not installed or require updating. To update, run:
        brew update
        brew uninstall --ignore-dependencies libimobiledevice
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    • ios-deploy 1.9.0
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] IntelliJ IDEA Community Edition (version 2017.1.4)
    • Flutter plugin version 14.0
    • Dart plugin version 171.4694.29

[✓] Connected devices
    • Pixel        • FA71A0302167                             • android-arm • Android 7.1.2 (API 25)
    •              • c0e109b7e55a826627aa0f3e79af8931534f3554 • ios         • iOS  ()
    • Posse iPhone • ac24c3de3cbe39efed542095769ce83fc89630f7 • ios         • iOS 10.3.1 (14E304)
    • iPhone 6     • 97A6FC25-AE09-4302-A8B3-565058389FE7     • ios         • iOS 10.3 (simulator)

```

I tried different combinations but this is what I see after pointing flutter config to an earlier version of gradle (3.3 - previously 3.5). The new exception is immediately below (file exception seems to be gone). I also included the outputs from flutter doctor and flutter run.

```
FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all files for configuration ':app:debugCompileClasspath'.
> Failed to transform file 'flutter-x86.jar' to match attributes {artifactType=android-classes} using transform JarTransform
   > Transform output file /Users/albertlardizabal/Posse/clients/test/test-flutter/build/app/intermediates/flutter/flutter-x86.jar does not exist.
```


```
☆ flutter doctor
[✓] Flutter (on Mac OS X 10.12.5 16F73, locale en-US, channel master)
    • Flutter at /Users/albertlardizabal/dev/flutter
    • Framework revision 37e32d5aee (20 hours ago), 2017-06-12 12:23:13 -0700
    • Engine revision ffe8181ffe
    • Tools Dart version 1.24.0-dev.6.7

[✓] Android toolchain - develop for Android devices (Android SDK 25.0.3)
    • Android SDK at /Users/albertlardizabal/Library/Android/sdk
    • Platform android-25, build-tools 25.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] iOS toolchain - develop for iOS devices (Xcode 8.3.3)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 8.3.3, Build version 8E3004b
    • ios-deploy 1.9.0
    • CocoaPods version 1.2.1

[✓] Android Studio (version 2.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Gradle version 3.2
    • Java version OpenJDK Runtime Environment (build 1.8.0_112-release-b06)

[✓] Gradle (3.3)
    • gradle-dir = /usr/local/Cellar/gradle/3.3
    • Flutter supports building with Gradle from Android Studio.
      Consider removing your gradle-dir setting by running:
      flutter config --gradle-dir=

[✓] IntelliJ IDEA Community Edition (version 2017.1.4)
    • Flutter plugin version 14.0
    • Dart plugin version 171.4694.29
```

```
☆ flutter run -d FA71A0302167
Launching lib/main.dart on Pixel in debug mode...
Running 'gradle assembleDebug'...                         
Configuration 'debugCompile' in project ':app' is deprecated. Use 'debugImplementation' instead.
Configuration 'profileCompile' in project ':app' is deprecated. Use 'profileImplementation' instead.
Configuration 'releaseCompile' in project ':app' is deprecated. Use 'releaseImplementation' instead.
Configuration 'compile' in project ':app' is deprecated. Use 'implementation' instead.
Configuration 'androidTestCompile' in project ':app' is deprecated. Use 'androidTestImplementation' instead.
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)
registerResGeneratingTask is deprecated, use registerGeneratedFolders(FileCollection)
Configuration 'compile' in project ':firebase_analytics' is deprecated. Use 'implementation' instead.
Configuration 'debugProvided' in project ':firebase_analytics' is deprecated. Use 'debugCompileOnly' instead.
Configuration 'releaseProvided' in project ':firebase_analytics' is deprecated. Use 'releaseCompileOnly' instead.
Configuration 'compile' in project ':firebase_database' is deprecated. Use 'implementation' instead.
Configuration 'debugProvided' in project ':firebase_database' is deprecated. Use 'debugCompileOnly' instead.
Configuration 'releaseProvided' in project ':firebase_database' is deprecated. Use 'releaseCompileOnly' instead.
Configuration 'debugProvided' in project ':path_provider' is deprecated. Use 'debugCompileOnly' instead.
Configuration 'releaseProvided' in project ':path_provider' is deprecated. Use 'releaseCompileOnly' instead.
Configuration 'compile' in project ':permit' is deprecated. Use 'implementation' instead.
Configuration 'debugProvided' in project ':permit' is deprecated. Use 'debugCompileOnly' instead.
Configuration 'releaseProvided' in project ':permit' is deprecated. Use 'releaseCompileOnly' instead.
Configuration 'debugProvided' in project ':url_launcher' is deprecated. Use 'debugCompileOnly' instead.
Configuration 'releaseProvided' in project ':url_launcher' is deprecated. Use 'releaseCompileOnly' instead.

FAILURE: Build failed with an exception.

* What went wrong:
Could not resolve all files for configuration ':app:debugCompileClasspath'.
> Failed to transform file 'flutter-x86.jar' to match attributes {artifactType=android-classes} using transform JarTransform
   > Transform output file /Users/albertlardizabal/Posse/clients/test/test-flutter/build/app/intermediates/flutter/flutter-x86.jar does not exist.
```

--- a/packages/flutter_tools/lib/src/android/gradle.dart
+++ b/packages/flutter_tools/lib/src/android/gradle.dart
@@ -211,6 +211,17 @@ Future<Null> buildGradleProjectV1(String gradle) async {
   printStatus('Built $gradleAppOutV1 (${getSizeAsMB(apkFile.lengthSync())}).');
 }
 
+File findApkFile(String buildDirectory, String buildModeName) {
+  final String apkFilename = 'app-$buildModeName.apk';
+  File apkFile = fs.file('$buildDirectory/$apkFilename');
+  if (apkFile.existsSync())
+    return apkFile;
+  apkFile = fs.file('$buildDirectory/$buildModeName/$apkFilename');
+  if (apkFile.existsSync())
+    return apkFile;
+  return null;
+}
+
 Future<Null> buildGradleProjectV2(String gradle, String buildModeName, String target, String kernelPath) async {
   final String assembleTask = "assemble${toTitleCase(buildModeName)}";
 
@@ -244,8 +255,9 @@ Future<Null> buildGradleProjectV2(String gradle, String buildModeName, String ta
     throwToolExit('Gradle build failed: $exitcode', exitCode: exitcode);
 
   final String buildDirectory = getGradleAppOutDirV2();
-  final String apkFilename = 'app-$buildModeName.apk';
-  final File apkFile = fs.file('$buildDirectory/$apkFilename');
+  final File apkFile = findApkFile(buildDirectory, buildModeName);
+  if (apkFile == null)
+    throwToolExit('Gradle build failed to produce an Android package.');
   // Copy the APK to app.apk, so `flutter run`, `flutter install`, etc. can find it.
   apkFile.copySync('$buildDirectory/app.apk');
 

--- a/packages/flutter_tools/test/android/gradle_test.dart
+++ b/packages/flutter_tools/test/android/gradle_test.dart
@@ -1,0 +1,34 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:file/file.dart';
+import 'package:file/memory.dart';
+import 'package:flutter_tools/src/android/gradle.dart';
+import 'package:test/test.dart';
+
+import '../src/context.dart';
+
+const String _kBuildDirectory = '/build/app/outputs';
+
+void main() {
+  FileSystem fs;
+
+  setUp(() {
+    fs = new MemoryFileSystem();
+    fs.directory('$_kBuildDirectory/release').createSync(recursive: true);
+    fs.file('$_kBuildDirectory/app-debug.apk').createSync();
+    fs.file('$_kBuildDirectory/release/app-release.apk').createSync();
+  });
+
+  group('gradle', () {
+    testUsingContext('findApkFile', () {
+      expect(findApkFile(_kBuildDirectory, 'debug').path,
+             '/build/app/outputs/app-debug.apk');
+      expect(findApkFile(_kBuildDirectory, 'release').path,
+             '/build/app/outputs/release/app-release.apk');
+    }, overrides: <Type, Generator>{
+      FileSystem: () => fs,
+    });
+  });
+}

